File: Data Structure/Python/double_linked_list.py
class DoubleLinkedList:
    class Node:
        def __init__(self, data = None, next = None, prev = None):
            self.data = data
            self.next = next
            self.prev = prev
            
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0
        
    def addToTail(self, value):
        self.length += 1
        if self.head == None:
            self.head = self.tail = self.Node(value, None, None)
            self.head.next = self.tail
            tail.next = None
            self.head.prev = None
        else:
            nd = self.Node(value, None, None)
            nd.prev = self.tail
            self.tail.next = nd
            self.tail = nd
            
    def addToHead(self, value):
        self.length += 1
        if self.head == None:
            self.head = self.tail = self.Node(value, None, None)
            self.head.next = self.tail
            self.tail.next = None
            self.head.prev = None
        else:
            nd = self.Node(value, None, None)
            nd.next = self.head
            self.head.prev = nd
            self.head = nd
    
    def insertInPos(self, index, value):
        if index < 0 or index > self.length + 1:
            return
        if index == 0: addToHead(value)
        if index == self.length + 1: addToTail(value)
        tmp = self.head
        nd = self.Node(value, None, None)
        n = 0
        while n!= index:
            tmp = tmp.next
            n += 1
        tmp.prev.next = nd
        nd.next = tmp
        nd.prev = tmp.prev
        tmp.prev = nd
        self.length += 1
            
    def getLength(self):
        return self.length
    
    def remove(self, value):
        curr = self.head
        while curr != None:
            if curr.data == value:
                tmp = curr
                if curr == self.tail:
                    self.tail = self.tail.prev
                    self.tail.next = None
                if curr.next != None:
                    curr.next.prev = curr.prev
                if curr.prev != None:
                    curr.prev.next = curr.next
                    curr = curr.next
                else:
                    self.head = curr.next
                    curr = self.head
            else:
                curr = curr.next
        self.length -= 1
        
    def search(self, value):
        curr = self.head
        index = 0
        while curr != None:
            if curr.data == value:
                return (True, index)
            curr = curr.next
            index += 1
        return (False, -1)

    def showFromHead(self):
        if self.head == None: return
        curr = self.head
        while curr != None:
            print(curr.data, sep=' ')
            curr = curr.next
            
    def showFromTail(self):
        if self.head == None: return
        curr = self.tail
        while curr != None:
            print(curr.data, sep=' ')
            curr = curr.prev     
            
    def __str__(self):
        if self.head == Nonde: return 'LinkedList []'
        curr = self.head
        out = 'LinkedList: '
        while curr != None:
            out += str(curr.data)+' '
            curr = curr.next
        return out  
    
    def sort(self):
        if self.head == None: return
        for i in range(0, self.length - 1):
            tmp = self.head
            while tmp.next != None:
                if tmp.data > tmp.next.data:
                    tmp.data, tmp.next.data = tmp.next.data, tmp.data
                tmp = tmp.next

File: Data Structure/C++/double_linked_list.cpp
#include <iostream>

template<typename T>
class doubleLinkedList
{
    private:
    struct node;
    node* head, *tail;
    size_t count;
    public:
    doubleLinkedList();
    ~doubleLinkedList();
    void addToHead(const T&);
    void addToTail(const T&);
    void showFromHead();
    void showFromTail();
    int search(const T&) const;
    void eraseElementByVal(const T&);
    void insert(const size_t&, const T&);
    void sort(const bool& ascending = true);
    
};


template<typename T>
struct doubleLinkedList<T>::node
{
    T data;
    node* next, *prev;
    node() = default;
    node(const T& val) : data(val), next(nullptr), prev(nullptr){}
};

template<typename T>
doubleLinkedList<T>::doubleLinkedList() : head(nullptr), tail(nullptr), count(0)
{
}  

template<typename T>
doubleLinkedList<T>::~doubleLinkedList()
{
    while(head != nullptr)
    {
        node* tmpNewHead = head->next;
        delete head;
        head = tmpNewHead;
    }
    tail = nullptr;
    count = 0;
}

template<typename T>
void doubleLinkedList<T>::addToHead(const T &val)
{
    node* newNode = new(std::nothrow) node(val);
    if(newNode != nullptr)
    {
        if (head == nullptr) 
        {
            head = tail = newNode;
            head -> next = tail;
            tail ->next = nullptr;
            head -> prev = nullptr;
        }
        else 
        {
            newNode -> next = head;
            head -> prev = newNode;
            head = newNode;
        }
    }
    ++count;
}

template<typename T>
void doubleLinkedList<T>::showFromHead()
{
    if(head == nullptr) return;
    node* tmp = head;
    while(tmp != nullptr)
    {
        std::cout<<tmp -> data<<" ";
        tmp = tmp -> next;
    }
    std::cout<<std::endl;
}

template<typename T>
void doubleLinkedList<T>::showFromTail()
{
    if(head == nullptr) return;
    node* tmp = tail;
    while(tmp != nullptr)
    {
        std::cout<<tmp -> data<<" ";
        tmp = tmp -> prev;
    }
    std::cout<<std::endl;
}

template<typename T>
void doubleLinkedList<T>::addToTail(const T &val)
{
    node* newNode = new(std::nothrow) node(val);
    if(newNode != nullptr)
    {
        if (head == nullptr) 
        {
            head = tail = newNode;
            head -> next = tail;
            tail ->next = nullptr;
            head -> prev = nullptr;
        }
        else 
        {
            newNode -> prev = tail;
            tail -> next = newNode;
            tail = newNode;
        }
    }
    ++count;
}

template<typename T>
int doubleLinkedList<T>::search(const T &val) const
{
    if(head == nullptr) return -1;
    node* tmp = head;
    size_t pos{0};
    while(tmp != nullptr)
    {
        if(tmp -> data == val) return pos;
        tmp = tmp -> next;
        ++pos;
    }
    return -1;
}   

template<typename T>
void doubleLinkedList<T>::eraseElementByVal(const T& val)
{
   for ( node *current = head; current != nullptr; )
   {
       if ( current-> data == val )
       {
           node *tmp = current;
           if( current == tail)
           {
               tail = tail -> prev;
               tail -> next = nullptr;
           }
           if ( current->next != nullptr )
           {
               current->next->prev = current->prev;
           }

           if ( current->prev != nullptr )
           {
               current -> prev -> next = current -> next;
               current = current -> next;
           }
           else
           {
               head = current -> next;
               current = head;
           }
           delete tmp;
           --count;
       }
       else
       {
           current = current->next;
       } 
    }
}


template<typename T>
void doubleLinkedList<T>::insert(const size_t& pos, const T& val)
{
    if( head == nullptr || pos < 0 || pos > count + 1) return;
    if(pos == 0) addToHead(val);
    if(pos == count + 1) addToTail(val);
    
    node* tmp = head;
    node* newNode = new node(val);
    size_t n{0};
    while(n != pos)
    {
        tmp = tmp -> next;
        ++n;
    }

    tmp -> prev -> next = newNode;
    newNode -> next = tmp;
    newNode -> prev = tmp -> prev;
    tmp -> prev = newNode;
    ++count;
    
}

template<typename T>
void doubleLinkedList<T>::sort(const bool& ascending)
{
     if (head == nullptr) return;

   for(int i = 0; i != count; i++)
   {  
        node *temp = head;
        while(temp -> next != nullptr) 
        {     
            if(temp -> data > temp -> next -> data) 
            {
                T tmpVal = temp -> data;     
                temp -> data = temp -> next -> data;
                temp -> next -> data = tmpVal;
            }
            temp = temp -> next;  
        }
    }
}

int main()
{
	
}

